### PR TITLE
Replace proc-macro-error with proc-macro2-diagnostics

### PR DESCRIPTION
**Pull request by tamird**
See individual commits.

I realize this crate is deprecated in favor of clap, but it still seems to
enjoy some 10,000-40,000 downloads per day over the past three months. Updating
to syn@2 is likely to have positive effect on compilation time for dependents.

See https://crates.io/crates/structopt-derive for download graphs.

### CODE REVIEW

**Comment by tamird** [comment]
@TeXitoi could you have a look please?

**Comment by TeXitoi** [comment]
That’s a big change for a frozen crate. I’ll try to have a look, but no promise.

File: tests/ui/raw.stderr [hunk 1]
```diff
@@ -1,18 +1,10 @@
 error: `#[structopt(raw(...))` attributes are removed in structopt 0.3, they are replaced with raw methods
-
-  = help: if you meant to call `clap::Arg::raw()` method you should use bool literal, like `raw(true)` or `raw(false)`
-  = note: if you need to call `clap::Arg/App::case_insensitive` method you can do it like this: #[structopt(case_insensitive = true)]
```
Comment (TeXitoi) [inline tests/ui/raw.stderr:4]: This is a regression

Comment (tamird) [reply tests/ui/raw.stderr:4]: Yes, and one that I've called out [here](https://github.com/TeXitoi/structopt/pull/536/files#diff-65b84fea55a38ced0e2d71af5d960af930f9cf65279f00411af3e36fe673332aR131).

**Comment by epage** [comment]
> I realize this crate is deprecated in favor of clap, but it still seems to
enjoy some 10,000-40,000 downloads per day over the past three months. Updating
to syn@2 is likely to have positive effect on compilation time for dependents.

imo that effort would be better spent migrating some of the more popular users.

**Comment by tamird** [comment]
You mean like this?
https://github.com/rust-lang/git2-rs/pull/1007
https://github.com/brendanzab/codespan/pull/357
https://github.com/marshallpierce/rust-base64/pull/262

File: tests/ui/raw.stderr (hunk 1 above)
Comment (jayvdb) [reply tests/ui/raw.stderr:4]: That regression should be noted in a CHANGELOG.md entry ?

Comment (tamird) [reply tests/ui/raw.stderr:4]: Would that unblock this PR?

Comment (TeXitoi) [reply tests/ui/raw.stderr:4]: No regression allowed for a crate in maintenance just to optimize compilation time. Better to migrate to clap if you want to improve compilation times.

**Comment by TeXitoi** [comment]
I close this MR as there is no activity.

**Comment by tamird** [comment]
If you can close it, maybe you can also review and accept it?

**Comment by TeXitoi** [comment]
Sorry. I don't want a regression, and there is one.

Also, I'm quite anxious at this PR as this is a big library change, and it might cause a behavior change.

I know that you worked in this and that closing this deny your work, but it adds more risk than advantages for my point of view.

**Comment by tamird** [comment]
Ack. Salvaged some useful stuff into #542 + added GitHub CI.

**Pull request closed without merging**